### PR TITLE
Update code.py

### PR DESCRIPTION
**Pull request by Shreyas0812**
I have added a small piece which will give us the length of lines, which will be useful while calculating the distance.

Also in initPop, I have created a population of 3x5, 3 - number of bots, 5 - no of lines, which will contain bot number assigned to the line as discussed earlier

### CODE REVIEW

File: code.py [hunk 1]
```diff
@@ -17,9 +17,23 @@ def __init__(self, numBots, lines, pose):
...
+			x2 = l[1][0]
+			y2 = l[1][1]
+			
+			#print (x1,y1,x2,y2)
+			dist = math.sqrt((x2 - x1)**2 + (y2 - y1)**2)
+			self.length.append(dist)
+			
 	# Initialize population
 	def initPop(self):
+		self.population = np.random.random_integers(low = 0, high = self.bots-1, size = (self.bots, self.numLines))
```
Comment (karpenet) [inline code.py]: Shouldn't size be (self.popSize, self.numLines)?

File: code.py [hunk 2]
```diff
@@ -17,9 +17,23 @@ def __init__(self, numBots, lines, pose):
...
+		#length of the lines
+		self.length = []
+		for l in lines:
+			x1 = l[0][0]
+			y1 = l[0][1]
+			x2 = l[1][0]
+			y2 = l[1][1]
+			
+			#print (x1,y1,x2,y2)
+			dist = math.sqrt((x2 - x1)**2 + (y2 - y1)**2)
```
Comment (karpenet) [inline code.py]: This is perfect, but np.linalg.norm might be easy to use

**Review by karpenet — CHANGES_REQUESTED** [review]
Please review comments.

File: code.py (hunk 2 above)
Comment (Shreyas0812) [reply code.py]: Oh okay. I'll change it in the next pull req

File: code.py (hunk 1 above)
Comment (Shreyas0812) [reply code.py]: Here we are defining the population right, so its basically the number of possible solutions we want. self.popSize is not being defined right, we are defining the population here. So what should I put it in its place.

Comment (karpenet) [reply code.py]: Your population consists of 'm' individuals/solutions, out of which you find the best one. Check the paper again.

Comment (Shreyas0812) [reply code.py]: Yeah, although it is something we define by ourselves and not as such dependent on any data, so what should i define it as?

Comment (Shreyas0812) [reply code.py]: I'll define it as 4 in the next pull request

File: code.py [hunk 3]
```diff
@@ -32,7 +34,8 @@ def __init__(self, numBots, lines, pose):
 			
 	# Initialize population
 	def initPop(self):
-		self.population = np.random.random_integers(low = 0, high = self.bots-1, size = (self.bots, self.numLines))
+		pop_size = 4
```
Comment (karpenet) [inline code.py]: Global variables to be defined in __init__(), but I'm merging this for now.

File: code.py [hunk 4]
```diff
@@ -53,12 +56,33 @@ def nextLine(self, pose):
 
 
 	# Function D in paper: returns dist b/w oldpose and next line
-	def nextDist(self, pose, line):
+	def nextDist(self, pose, line, index_bot):
+		x0 = pose[0]
+		x1 = pose[1]
```
Comment (karpenet) [inline code.py]: y0?

File: code.py [hunk 5]
```diff
@@ -53,12 +56,33 @@ def nextLine(self, pose):
...
+		x1 = line[0][0]
+		y1 = line[0][1]
+		x2 = line[1][0]
+		y2 = line[1][1]
+		
+		d1 = math.sqrt((x1 - x0)**2 + (y1 - y0)**2)
+		d2 = math.sqrt((x2 - x0)**2 + (y2 - y0)**2)
+		
+		if d1 < d2:
+			self.pose[index_bot] = [x2, y2]
```
Comment (karpenet) [inline code.py]: You can return d1 here rather than finding min() again later.

File: code.py [hunk 6]
```diff
@@ -53,12 +56,33 @@ def nextLine(self, pose):
...
+		else:
+			self.pose[index_bot] = [x1, y1]
+		dist = min(d1,d2)
 		return dist
 
 
 	# Heuristics func in paper: Return distance/cost of a single input group
-	def groupDist(self, group):
```
Comment (karpenet) [inline code.py:47]: This is the heuristic() function in the paper. It takes a group as an input, and orders the lines in the group, not an individual.

**Review by karpenet — CHANGES_REQUESTED** [review]
Check the heuristics function again.

File: code.py (hunk 4 above)
Comment (Shreyas0812) [reply code.py]: Oh, I am sorry, I must've typed it wrong. rest assured , I did it correctly in my machine and the code is working

File: code.py (hunk 5 above)
Comment (Shreyas0812) [reply code.py]: Yeah,... will do

File: code.py (hunk 6 above)
Comment (Shreyas0812) [reply code.py:47]: oh okay. Isn't the group referring to one of the solution because it returns cost of that. I will go through the paper as well

Comment (Shreyas0812) [reply code.py:47]: Here the cost will return the total distance traveled by all bots in that particular solution in the population.

Comment (karpenet) [reply code.py:47]: Groups refers to list of lines for the i-th robot. For ex. group_1 is a set of line that robot 1 prints. Hence, you find robot 1's cost.

Comment (Shreyas0812) [reply code.py:47]: Oh okay. I get what needs to be done. 
But bhaiya, I have a doubt. 
Getting individual cost of each robot in the solution and then trying to make them all as close to each other as possible.
And Getting the total cost of the solution and minimizing it as a whole.

Won't they both result in the same thing. I get the point, will work on it as well.
I am really sorry if I sound rude or anything.

Comment (Shreyas0812) [reply code.py:47]: In this, c[i] will store the array of costs of the respective bots, and the max of them will be returned. Please tell me if this is correct or do I need to make changes.

File: code.py [hunk 7]
```diff
@@ -69,20 +69,25 @@ def nextDist(self, pose, line, index_bot):
...
+	def groupDist(self, group):
 		index_line = 0
-		cost = 0
-		for index_bot in sol:
+		c = np.empty(len(self.pose))
+		for i in range(len(c)):
+			c[i] = 0
+			
+		for index_bot in group:
 			dist = self.nextDist(pose = self.pose[index_bot], line = self.lines[index_line], index_bot = index_bot)
```
Comment (karpenet) [inline code.py]: nextDist() should be called with the next "nearest" line. Not the next line in the list.

File: code.py [hunk 8]
```diff
@@ -69,20 +69,25 @@ def nextDist(self, pose, line, index_bot):
...
-			cost = cost + dist + self.length[index_line]
-			index_line = index_line + 1		
+			c[index_bot] = c[index_bot] + dist + self.length[index_line]
+			index_line = index_line + 1
+			
+		cost = max(c)
```
Comment (karpenet) [inline code.py]: One group will have only one cost. max() function is hence redundant here.

File: code.py [hunk 9]
```diff
@@ -69,20 +69,25 @@ def nextDist(self, pose, line, index_bot):
...
+			return d2
 
 
 	# Heuristics func in paper: Return distance/cost of a single input group
-	def groupDist(self, sol):
+	def groupDist(self, group):
 		index_line = 0
-		cost = 0
-		for index_bot in sol:
+		c = np.empty(len(self.pose))
```
Comment (karpenet) [inline code.py]: It's easier to use:
`c = [0 for i in range(len(self.pose))]`

**Review by karpenet — CHANGES_REQUESTED** [review]
Group logic seems incorrect.

File: code.py (hunk 9 above)
Comment (Shreyas0812) [reply code.py]: Oh, yeah. Will make the change

File: code.py (hunk 7 above)
Comment (Shreyas0812) [reply code.py]: Oh okay, got it. I'll try to make the changes

File: code.py (hunk 8 above)
Comment (Shreyas0812) [reply code.py]: Each group will have 3 costs for each bot. Cost is just returning the max out of those 3 values

Comment (karpenet) [reply code.py]: Each individual should have 3 groups, each group should have one cost.

Comment (Shreyas0812) [reply code.py]: Yeah. I have read and understood the paper. By tomorrow, I will upload a draft of the code as given in the paper

**Pull request merged**